core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -101,7 +102,7 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         FailedRerouteAllocation allocation = new FailedRerouteAllocation(allocationDeciders, routingNodes, clusterState.nodes(), failedShards, clusterInfoService.getClusterInfo());
         boolean changed = false;
         for (FailedRerouteAllocation.FailedShard failedShard : failedShards) {
-            changed |= applyFailedShard(allocation, failedShard.shard, true, new UnassignedInfo(UnassignedInfo.Reason.ALLOCATION_FAILED, failedShard.details));
+            changed |= applyFailedShard(allocation, failedShard.shard, true, new UnassignedInfo(UnassignedInfo.Reason.ALLOCATION_FAILED, failedShard.message == null ? null : failedShard.message + (failedShard.failure == null ? "" : ", failure: " + ExceptionsHelper.detailedMessage(failedShard.failure))));
```

Comment:
Maybe include the exception in `UnassignedInfo` instead of a detailed message? So that the entire exception is included once the cluster state is serialized, like we do on other places to now and it is useful too.